firedrake/functionspaceimpl.py

```diff
@@ -514,6 +514,9 @@ def boundary_nodes(self, sub_domain, method):
         """
         return self._shared_data.boundary_nodes(self, sub_domain, method)
 
+    def collapse(self):
+        return FunctionSpace(self.mesh(), self.ufl_element())
```

Comment:
This is wrong, you need:

```python
def collapse(self):
   from firedrake import FunctionSpace
   return FunctionSpace(self.mesh(), self.ufl_element())
```

This is so you get a `WithGeometry` appropriately.

Reply:
Gotcha. Thanks, updated below

Reply:
I presume this still works for the notebook. Can you add a simple test that this does the "correct" thing?

That is, collapsing subspaces should return something that is `==` to just making the subspace "directly"

Reply:
The notebook works on my machine with this, yes.

Will write a test now

Reply:
Maybe I misunderstood something, but my simple test below fails, with
```
    def test_function_space_collapse(cg1):
>       assert cg1 == cg1.collapse()
E       assert WithGeometry(...), dim=2), 1)) == FunctionSpace(...1), name=None)
```

Reply:
Ah, that means, I think you need to implement `collapse` in `WithGeometry` as:

```python
def collapse(self):
    return type(self)(self.topological.collapse(), self.mesh())
```

Reply:
Cheers Lawrence, test passes now